docs/proposals/keeping-transient-pod-name-spaces-alive.md

```diff
@@ -0,0 +1,54 @@
+# Keeping transient pod name spaces alive
+
+## Introduction
+The goal of this proposal is to give administrators options to keep namespaces and therefore APB pods after the execution. The use cases for this feature are demos and debugging APBs.
+
+## Problem Description
+The [bug](https://bugzilla.redhat.com/show_bug.cgi?id=1497766) was created when we moved to creating transient namespaces during execution of the APB pod. What happens  This bug creates issues for debugging APB's as well as issues with demos. 
+
+## Implementation Details.
+
+### Configuration Values
+
+* `keep_namespace` or `ClusterConfig.KeepNamespace` - Parameter to always keep name space no matter what.
+* `keep_namespace_on_error` or `ClusterConfig.KeepNamespaceOnError` -  parameter to keep name space around if an error occurs in the play book that is running.
+
+Example:
+```yaml
+...
+openshift:
+  ....
+  keep_namespace: false
+  keep_namespace_on_error: true
+...
+```
+
+**NOTE: Both will default to false in openshift-ansible, but will be set to true for keep_namespace_on_error**
+
+### Major Code Change
+
+We will use the [DestroySandbox](https://github.com/openshift/ansible-service-broker/blob/34f643eec5349f58300e4e802581a65f4120976c/pkg/apb/svc_acct.go#L225) method to determine if we should delete the sandbox. This method is used already used by all methods that run the APB.
+
+Here we can [get the pod](https://godoc.org/k8s.io/client-go/kubernetes/typed/core/v1#PodInterface). Then we can use the [PodPhase](https://godoc.org/k8s.io/api/core/v1#PodStatus) from the [PodSatus](https://godoc.org/k8s.io/api/core/v1#Pod). The [PodPhase](https://godoc.org/k8s.io/api/core/v1#PodPhase) will be in error if it is `PodFailed` or `PodUnknown`. 
+
+
+The logic will be in the `pkg/apb/svc_account.go` file in the `DestroySandbox` method:
+```golang
+....
+pod, err := k8scli.CoreV1().Pods(executionContext.Namespace).Get(executionContext.PodName, metav1.GetOptions{})
+if err != nil {
+        s.log.Errorf("Unable to retrieve pod - %v", err)     
+}
+if !brokerConfig.keepNamespace || !((pod.Status.Phase == apiv1.PodFailed || pod.Status.Phase == apiv1.PodUnknown || err != nil) && brokerConfig.keepNamespaceOnError) {
+    ... Delete Namespace 
+}
+...Delete role bindings.
```

Comment:
So the only thing(s) that should survive when we are keeping the namespace is 1) the namespace and 2) the pod? I imagine we would want to remove everything else.

Reply:
If we are configured to keep the namespace, I assume we want to keep everything in it we created to aid in debugging.

Reply:
Are we concerned about leaking role related resources like SA's and RoleBindings?

Reply:
We are removing the role bindings, the thinking is that you can then give other users access to the new namespace, they can debug the APB without getting elevated permissions into the target namespace. 

@rthallisey 
> How much security are we gaining by having an ephemeral namespace versus a locked down persistent namespace for APBs?

The above is the added security that we will get.

Reply:
> We are removing the role bindings, the thinking is that you can then give other users access to the 
> new namespace, they can debug the APB without getting elevated permissions into the target 
> namespace.
>
>>    How much security are we gaining by having an ephemeral namespace versus a locked
>> down persistent namespace for APBs?
>>
> The above is the added security that we will get.
>
I disagree. The namespace is not the source for the vulnerability because by itself it's not dangerous. The vulnerability is with the rolebindings for the namespace. If we manage the rolebindings on every APB namespace, then I don't think there's any harm in keeping them around for at least a period of time.  This will allow folks debug a failure in their APBs or to read the output logs from a successful run.  I would much rather have a namespace deletion timer than having to check for APB pod failures in our code and require another config setting in the broker to enable persistent namespaces.

Reply:
I don't disagree that the watcher that will delete them after a certain time is a good idea. We have captured that, and I think is just too much to implement in a bug fix. 

I have created a card on the backlog already:
https://trello.com/c/i5gUkUv4